subsys/bluetooth/audio/has.c

```diff
@@ -34,7 +56,362 @@ static ssize_t read_features(struct bt_conn *conn, const struct bt_gatt_attr *at
+#else
+	const char *name;
+#endif /* CONFIG_BT_HAS_PRESET_NAME_DYNAMIC */
+} has_preset_list[CONFIG_BT_HAS_PRESET_COUNT];
+
+static void process_control_point_work(struct k_work *work);
+
+static struct has_client *client_get(struct bt_conn *conn)
+{
+	struct has_client *client = NULL;
+
+	for (size_t i = 0; i < ARRAY_SIZE(has_client_list); i++) {
+		if (conn == has_client_list[i].conn) {
+			return &has_client_list[i];
+		}
+
+		/* first free slot */
+		if (!client && !has_client_list[i].conn) {
+			client = &has_client_list[i];
+		}
+	}
+
+	__ASSERT(client, "failed to get client for conn %p", (void *)conn);
+
+	client->conn = bt_conn_ref(conn);
+
+	k_work_init(&client->control_point_work, process_control_point_work);
+
+	return client;
+}
+
+static bool read_presets_req_is_pending(struct has_client *client)
+{
+	return client->read_presets_req.num_presets > 0;
+}
+
+static void read_presets_req_free(struct has_client *client)
+{
+	client->read_presets_req.num_presets = 0;
+}
+
+static void client_free(struct has_client *client)
+{
+	(void)k_work_cancel(&client->control_point_work);
+
+	read_presets_req_free(client);
+
+	bt_conn_unref(client->conn);
+
+	client->conn = NULL;
+}
+
+static struct has_client *client_find(struct bt_conn *conn)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(has_client_list); i++) {
+		if (conn == has_client_list[i].conn) {
+			return &has_client_list[i];
+		}
+	}
+
+	return NULL;
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	struct has_client *client;
+
+	BT_DBG("conn %p reason %d", (void *)conn, reason);
+
+	client = client_find(conn);
+	if (client) {
+		client_free(client);
+	}
+}
+
+BT_CONN_CB_DEFINE(conn_cb) = {
+	.disconnected = disconnected,
+};
+
+typedef uint8_t (*preset_func_t)(const struct bt_has_preset *preset, void *user_data);
+
+static void preset_foreach(uint8_t start_index, uint8_t end_index, preset_func_t func,
+			   void *user_data)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(has_preset_list); i++) {
+		const struct bt_has_preset *preset = &has_preset_list[i];
+
+		/* Skip ahead if out of range */
+		if (preset->index < start_index  || preset->index > end_index) {
+			continue;
+		}
+
+		if (func(preset, user_data) == BT_HAS_PRESET_ITER_STOP) {
+			return;
+		}
+	}
+}
+
+static uint8_t preset_found(const struct bt_has_preset *preset, void *user_data)
+{
+	const struct bt_has_preset **found = user_data;
+
+	*found = preset;
+
+	return BT_HAS_PRESET_ITER_STOP;
+}
+
+static struct bt_has_preset *preset_alloc(uint8_t index, enum bt_has_properties properties,
+					  const char *name)
+{
+	struct bt_has_preset *preset = NULL;
+
+	preset_foreach(BT_HAS_PRESET_INDEX_NONE, BT_HAS_PRESET_INDEX_NONE, preset_found, &preset);
+
+	if (preset != NULL) {
+		preset->index = index;
+		preset->properties = properties;
+#if defined(CONFIG_BT_HAS_PRESET_NAME_DYNAMIC)
+		utf8_lcpy(preset->name, name, ARRAY_SIZE(preset->name));
+		preset->name[strlen(name)] = '\0';
```

Comment:
`utf8_lcpy` ensures that the string is null terminated
```suggestion
		utf8_lcpy(preset->name, name, ARRAY_SIZE(preset->name));
```